[Reviewer: Chris] Resend status of alarms to alarm handler- cpp-common

AME2 opened the pull request:
The main changes here involve changes made to the alarm class hierarchy and creating functions to reraise the alarms every 30 seconds.

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+  
+  /// Queues a request to generate an alarm state change corresponding to the
+  /// CLEARED severity.
+  virtual void clear();
```
As the code stands, this doesn't need to be virtual as you don't override it in subclasses.

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+
+  /// Uses the _last_state_raised member variable to re-raise the latest state
+  /// of the alarm.
+  virtual void reraise_last_state();
```
Ditto

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+{
+public:
+  static AlarmManager& get_instance() {return _instance;}
+  static AlarmManager _instance;
```
This should be private - the instance must never be assigned to from outside the class, and must only be accessed through get_instance.

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+  static AlarmManager& get_instance() {return _instance;}
+  static AlarmManager _instance;
+  bool _terminated;
+  void add_alarm_to_list(BaseAlarm* alarm); 
```
I think 'register_alarm' is a better name as it describes what the function is doing in terms of purpose rather than technical details.

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+  // every 30 seconds. For each alarm it calls the reraise_last_state method.
+  virtual void reraise_alarms();
+  // This is used for storing all of the BaseAlarm objects as they get
+  // constructed.
```
Constructed or registered? (See above)

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
 /// @class Alarm
 ///
-/// Encapsulates an alarm active state and its associated alarm clear state.
+/// Encapsulates an alarm's only active state with its associated alarm clear state.
```
It's good to do this kind of touch-up work when you are in the area.

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -89,28 +155,51 @@ class Alarm
   virtual ~Alarm() {}
```
You are using 'virtual' a lot, but it's only necessary when you define a method which needs to be overridden by a subclass. Alarm and MultiSeverityAlarm are not subclassed, so they don't need any virtual methods.

bossmc replied:
Although, if you want to use polymorphism (treating a subclass as a superclass) you always want the destructor to be virtual (I think g++ warns about this, clang certainly does).

On the other hand, `virtual` is inherited from the parent class (in this case `Alarm`) and thus doesn't need to be specified here again.

AME2 replied:
I think the reason behind these virtual functions was so that we could have a MockAlarm subclass that could define mock methods to override these functions:
https://github.com/Metaswitch/cpp-common/blob/ReraisingAlarms/test_utils/mockalarm.h

chris-elford-metaswitch replied:
Makes sense.

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -89,28 +155,51 @@ class Alarm
+  
+  /// Indicates whether the alarm state currently maintained by this object
+  /// corresponds to the non-CLEARED severity.
+  virtual bool alarmed() {return _alarmed.load();}
```
You've defined the same method twice! You should consolidate this into the superclass - one of the main reasons for using inheritance is to re-use code.

chris-elford-metaswitch replied:
More to the point, is this function even used?

AME2 replied:
Yes- by the Cedar communication monitor. I'll consolidate them into the superclass.

chris-elford-metaswitch commented on src/alarm.cpp:
```diff
@@ -76,36 +77,145 @@ void AlarmState::clear_all(const std::string& issuer)
   TRC_DEBUG("%s cleared its alarms", issuer.c_str());
 }
 
+BaseAlarm::BaseAlarm(const std::string& issuer,
```
It's worth thinking a little about how you lay out the file. I'd define all of the methods for one class, then move on to the next.

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+  virtual void reraise_alarms();
+  // This is used for storing all of the BaseAlarm objects as they get
+  // constructed.
+  std::vector<BaseAlarm*> _global_alarm_list;
```
This is called _global_alarm_list, but it's not a global variable.

chris-elford-metaswitch commented on src/alarm.cpp:
```diff
@@ -76,36 +77,145 @@ void AlarmState::clear_all(const std::string& issuer)
+{
+  while (!_terminated)
+  {
+    for (std::vector<BaseAlarm*>::iterator it = _global_alarm_list.begin(); it != _global_alarm_list.end(); it++)
```
I'd prefer to use a [range loop](http://en.cppreference.com/w/cpp/language/range-for) here, but let's hold off until after Alex's write-up.

chris-elford-metaswitch commented on src/alarm.cpp:
```diff
@@ -76,36 +77,145 @@ void AlarmState::clear_all(const std::string& issuer)
+    }
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    ts.tv_sec += 30;
```
We discussed this timing issue before when we went over your design. You are waiting 30s *after* re-syncing all alarms, but the re-sync will take non-zero time. This means that you don't re-sync every 30s.

I think what you should do is:
- store the previous time on the AlarmManager.
- add 30s to it and subtract the current time.
- if that time is more than 0, wait for it.
- otherwise, write a log and sleep until the next valid 30s interval.

chris-elford-metaswitch commented on src/alarm.cpp:
```diff
@@ -76,36 +77,145 @@ void AlarmState::clear_all(const std::string& issuer)
+AlarmManager::AlarmManager()
+{
+  _terminated = false;
+  pthread_mutex_init(&_lock, NULL);
```
I don't quite understand what this code is doing. A short high-level comment might help e.g. "Protect x with a condition variable."

chris-elford-metaswitch commented on src/alarm.cpp:
```diff
@@ -76,36 +77,145 @@ void AlarmState::clear_all(const std::string& issuer)
+  _last_state_raised->issue();
+}
+
+void* AlarmManager::reraise_alarms_function(void* data)
```
Add a comment explaining why this function is defined with this signature. I think it needs to fit the signature expected by pthread_create.
- What is the data passed in?
- Why do you return NULL?

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+  // constructed.
+  std::vector<BaseAlarm*> _global_alarm_list;
+  pthread_mutex_t _lock;
+  pthread_cond_t _terminating_variable;
```
I'm not sure this name conveys what the condition variable is for - it's used when the thread is not terminating. I might be tempted just to call it _condition.

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+  // This is used for storing all of the BaseAlarm objects as they get
+  // constructed.
+  std::vector<BaseAlarm*> _global_alarm_list;
+  pthread_mutex_t _lock;
```
What does this lock protect? Please add a comment.

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+  void add_alarm_to_list(BaseAlarm* alarm); 
+  // Static function called by the reraising alarms thread. This simply calls
+  // the 'reraise_alarms' member method.
+  static void* reraise_alarms_function(void* data); 
```
I assumed that this would be private. Is that possible? (My C++ isn't quite good enough to tell.)

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+class BaseAlarm
+{
+public:
+  BaseAlarm(const std::string& issuer,
```
This should be protected - you don't want people creating BaseAlarm objects.

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -74,12 +74,78 @@ class AlarmState
+  AlarmState _clear_state;
+  
+  /// Keeps track of whether the alarm is raised
+  std::atomic<bool> _alarmed;
```
I'm not convinced by this bool. You are currently exposed to the following threading bug:
1. Thread A sets _alarmed to true.
2. Thread B sets _alarmed to false.
3. Thread B sets _last_state_raised to the cleared state.
4. Thread A sets _last_state_raised to the raised state.
At this point, _last_state_raised is the raised state, but _alarmed is false!

At the same time, _last_state_raised contains all of the information about whether the alarm is raised or not - simply compare it to the cleared state.

I think there is an atomic intptr type which you can use to fix this bug.


All of the information about whether this is alarmed is stored in the _last_state_raised pointer.

chris-elford-metaswitch replied:
Please address this comment.

AME2 replied:
We no longer raise alarms based on the _alarmed variable. We now check the new _last_state_raised variable and if this is different from the state of the alarm we which to raise, we raise it.
I believe the only place that still checks this _alarmed variable is the Cedar communication monitor.

chris-elford-metaswitch replied:
As discussed, this needs fixing.

chris-elford-metaswitch replied:
You still haven't fixed this, or explained why it's OK.

chris-elford-metaswitch commented on src/alarm.cpp:
```diff
@@ -76,36 +77,145 @@ void AlarmState::clear_all(const std::string& issuer)
+
+void AlarmManager::reraise_alarms()
+{
+  while (!_terminated)
```
I'd add a debug level log here.

chris-elford-metaswitch commented on src/alarm.cpp:
```diff
@@ -76,36 +77,145 @@ void AlarmState::clear_all(const std::string& issuer)
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    ts.tv_sec += 30;
+    pthread_cond_timedwait(&_terminating_variable, &_lock, &ts);
+  }
```
I'd add an info level log here saying that the re-raise thread is terminating.

chris-elford-metaswitch commented on src/alarm.cpp:
```diff
@@ -76,36 +77,145 @@ void AlarmState::clear_all(const std::string& issuer)
+  pthread_mutex_init(&_lock, NULL);
+  pthread_condattr_t cond_attr;
+  pthread_condattr_init(&cond_attr);
+  pthread_condattr_setclock(&cond_attr, CLOCK_MONOTONIC);
```
Why is this necessary?

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -89,28 +155,51 @@ class Alarm
+///
+/// Encapsulates an alarm's two or more active states with its associated clear
+/// state. Used to manage the reporting of a fault condition, and subsequent
+/// clear of said condition.
```
Please add a comment explaining that this must be subclassed, and providing an example. Otherwise, someone will just come along and make all of the set methods public...

chris-elford-metaswitch commented on include/alarm.h:
```diff
@@ -89,28 +155,51 @@ class Alarm
+  virtual bool alarmed() {return _alarmed.load();}
+  
+protected:
+  /// These raise the alarm with the specified severity.
```
It's worth explaining again why these are protected.

chris-elford-metaswitch commented:
This is mostly good, but a few tweaks are needed. I'd like to see again before you merge.

chris-elford-metaswitch commented on src/alarm.cpp:
```diff
@@ -76,36 +77,145 @@ void AlarmState::clear_all(const std::string& issuer)
+{
+  _alarmed.exchange(true);
+  _last_state_raised = &_warning_state;
+  _warning_state.issue();
```
You've not added the same logic here as in the Alarm to only send if there is a change.

AME2 replied:
Yes, that logic said only clear an alarm if it is raised and (importantly) only raise an alarm if it is cleared. So I couldn't repeat that logic because with MultiStateAlarms we may want to raise them again if they are already raised. I'm going to change both bits of logic to base the decision off the _last_state_raised BaseAlarm member variable. If the last state raised of the alarm is different from the current state we raise the alarm.

chris-elford-metaswitch replied:
Sounds good to me!

chris-elford-metaswitch commented:
I've added a few more comments as I think there are still some bugs here. I'm also interested in how many blank lines you've deleted. We tend to use these to separate sections of code and make it more readable so it's not always right to remove them.

chris-elford-metaswitch commented:
Mostly good; just a couple of comments.

You've still not addressed my comment about the _alarmed flag. Last I remember, you were going to investigate fixing the threading issue as a way of ensuring that the interface exposed by this code does what is expected. Even if you think this doesn't need fixing, you should add a comment to explain why it doesn't.

AME2 commented:
I've removed the _alarmed private member function of the class BaseAlarm and have modified the alarmed() public method. This method now looks at the _last_state_raised private member variable to decide whether an alarm is currently active (in a non-CLEARED severity) or not. Then to stop the threading race condition I've included a mutex lock in my set_x methods.
In class BaseAlarm I've also moved the _first_alarm_raised variable into the private namespace and now use the methods start_resending_alarms and has_alarm_been_raised to write and read this variable.

johadalin commented:
I can't see any big problems in the last couple of commits. 
@rkd-msw , if you have the time, can you take a closer look at this. I don't have a particularly good view of the alarms code, so i may well be missing things.

rkd-msw commented:
LGTM.
